include/merlin/core_kernels.cuh

```diff
@@ -942,8 +942,9 @@ struct SelectUpsertKernelWithIO {
  */
 template <class K, class V, class M, uint32_t TILE_SIZE = 4>
 __global__ void upsert_kernel(const Table<K, V, M>* __restrict table,
-                              const K* __restrict keys, V** __restrict vectors,
-                              const M* __restrict metas,
+                              const K* __restrict keys,
```

Comment:
Hi @bashimao , could you move the changes that have no relationship with GDS API to a separate PR? I think that can help the GDS team review your design more efficiently, thanks!

Reply:
Done!